### ISSUE TITLE
Pod identity propagation delay to remote IP cache when using egress gateways

### ISSUE THREAD

**Issue by Champ-Goblem** [cilium/cilium#23976]
We seem to be experiencing some problems with Cilium Endpoint propagation on a GKE cluster with 2600 pods, 17 nodes and 80 cronjobs. Short-lived pods either created manually or via a cronjob sometimes experience connectivity problems when connecting to other internal services protected by network policies.

The failures often happen when there is a high pod churn on the cluster, this might be due to a number of cronjobs triggering at once or pods being deleted/created at the time the affected pod spawns.

The large number of pod updates seem to cause propagation delays of new pod IPs, which leads to the cilium IP cache not being updated in time on the remote nodes. If the delay is long enough it seems that the cache is not updated before the initial network connection is attempted, this leads to cilium classifying the identity of the cronjob pod as `reserved:world` which then causes the connection to be blocked by the network policy in that namespace. For example from cilium monitor logs during a failure event:

```
Policy verdict log: flow 0x0 local EP ID 902, remote ID world, proto 6, ingress, action deny, match none, 10.4.36.247:52942 -> 10.4.58.66:3306 tcp SYN
xx drop (Policy denied) flow 0x0 to endpoint 902, file bpf_lxc.c line 2003, , identity world->48169: 10.4.36.247:52942 -> 10.4.58.66:3306 tcp SYN
```

vs when the cronjob connects as expected:

```jsx
Policy verdict log: flow 0x0 local EP ID 902, remote ID 35916, proto 6, ingress, action allow, match L3-Only, 10.4.31.130:43190 -> 10.4.58.66:3306 tcp SYN
```

The network policy in that namespace looks like so:

```jsx
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: ns-block
  namespace: ns-a
spec:
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          netpolicylabel: ns-a
  - from:
    - namespaceSelector:
        matchLabels:
          name: istio
  - from:
    - namespaceSelector:
        matchLabels:
          name: forwarding
  - from:
    - namespaceSelector:
        matchLabels:
          name: logging
  podSelector: {}
  policyTypes:
  - Ingress
status: {}
```

I wrote a quick tool to watch the kubernetes event stream for cilium endpoint resource changes and then run `cilium bpf ipcache get <ip-addr>` on each node in the cluster to determine how long the propagation takes. For the majority of the time the delay is below 1 second, but in the event of 30/50/100 cilium endpoint updates the propagation delays can become in excess of 10s, leading to the issue.

We recently upgraded cilium to 1.13.0 and also reduced the identity labels at the same time as this upgrade. This has improved the situation and the problem occurs now less frequently, but it is still something which we would prefer to avoid happening at all.

Since the upgrade I have noticed that the failure timestamps for the jobs correspond with a large increase of:

`2023-02-21T12:30:06.348176806Z stderr F level=info msg="Trace[1237947675]: \"DeltaFIFO Pop Process\" ID:ns-a/test-db-swzfx,Depth:15,Reason:slow event handlers blocking the queue (21-Feb-2023 12:30:05.768) (total time: 579ms):" subsys=klog`

log lines, which is demonstrated in the below image:

![image](https://user-images.githubusercontent.com/19267423/220956939-e574d2a4-a184-4d66-9ff8-2522277e682e.png)

The green line is the rate of the `slow event handlers blocking the queue` logs and the yellow line corresponds to the creation/deletion logs of some of the recently failed jobs.

I believe the slow event handlers are down to the time it takes to process the updates in the cilium egress gateway related code. At the time of the second set of failures (the rightmost red highlight on the graph) a flamegraph of the cilium agent process shows a lot of time spent in `egressgateway.(*Manager).OnUpdateEndpoint` and `egressgateway.(*Manager).OnDeleteEndpoint` functions.

![image](https://user-images.githubusercontent.com/19267423/220957046-0ce74402-e7b4-4ec5-ad51-990f6ddb1abf.png)

Here is the left side of the graph following a cilium endpoint `OnAdd`:

![image](https://user-images.githubusercontent.com/19267423/220957070-022ff5c6-b402-4b87-8c3f-b35da6038399.png)

and the right side following a cilium endpoint `OnDelete`:

![image](https://user-images.githubusercontent.com/19267423/220957121-f07fa372-7a71-43bc-9171-01f011799bcd.png)

Regarding the egress gateway setup, we have two cilium egress gateway entries on the cluster which are responsible for SNATing the connections for ~40 pods, with one node acting as the designated gateway. The designated node runs no workloads apart from the daemonsets on the clusters, the cilium daemonset and any kubernetes components.

**Comment by Champ-Goblem** [comment]
This is still relevant

**Comment by pchaigno** [comment]
Did you try with the latest v1.13 release? We've improved this egress gateway logic recently. Some of the fixes made it into v1.13.2 and some will be in the v1.14.0 release.

**Comment by marseel** [comment]
Could you take a look at CPU usage for cilium-agent and also node that it's running on, please? This is quite a high-pod density setup and I am wondering if it might also have some impact on performance of cilium-agent.
